contracts/Penalizer.sol

```diff
@@ -23,14 +25,34 @@ contract Penalizer is IPenalizer{
         transaction.value,
         transaction.data) = RLPReader.decodeTransaction(rawTransaction);
         return transaction;
+    }
+
+    mapping(bytes32 => uint) public commits;
 
+    event Commit(bytes32 indexed commitHash, address sender);
+
+    /**
+     * any sender can call "commit(keccak(encodedPenalizeFunction))", to make sure
+     * no-one can front-run it to claim this penalization
+     */
+    function commit(bytes32 commitHash) external override {
+        commits[commitHash] = block.number;
+        emit Commit(commitHash, msg.sender);
     }
 
     modifier relayManagerOnly(IRelayHub hub) {
         require(hub.isRelayManagerStaked(msg.sender), "Unknown relay manager");
         _;
     }
 
+    modifier commitRevealOnly() {
+        bytes32 commitHash = keccak256(abi.encodePacked(keccak256(msg.data), msg.sender));
+        uint commitBlockNumber = commits[commitHash];
+        require(commitBlockNumber != 0, "no commit");
+        require(commitBlockNumber + PENALIZE_BLOCK_DELAY < block.number, "reveal penalize too soon");
+        _;
+    }
+
     function penalizeRepeatedNonce(
```

Comment:
Suggest original path to be removed. 
Registered relays can be added just to front-run penalization, decreasing motivation to monitor chain.